sql: avoid tenant ID reuse

knz opened the pull request:
Prior to this patch, `DROP TENANT` would delete rows from `system.tenants` and `CREATE TENANT` would reuse IDs not already in the table.

Tenant ID reuse is known to incur correctness problems in the following two cases:

- we do not yet wait for server shutdown when the service mode is changed to 'NONE'. This makes it possible to start dropping a tenant and then reusing its ID before servers from the previous record have fully shut down. This creates a possibility for a past server to start serving traffic for a freshly created tenant, which is unacceptable.

- we have a cache of tenant capabilites on every node. We do not yet implement an invalidation protocol to clear that cache upon tenant deletion. This makes it possible for requests from a new tenant to be authorized using the capabilities of a previous tenant, which is also unacceptable.

Until we implement the necessary synchronization, we can avoid the correctness issues by preventing ID reuse.

This patch achieves this -- by preventing the deletion of rows in system.tenants.

Release note: None
Epic: CRDB-23559
Fixes #100615.

knz committed:
sql: avoid tenant ID reuse

Prior to this patch, `DROP TENANT` would delete rows from
`system.tenants` and `CREATE TENANT` would reuse IDs not already in
the table.

Tenant ID reuse is known to incur correctness problems in the
following two cases:

- we do not yet wait for server shutdown when the service mode
  is changed to 'NONE'. This makes it possible to start
  dropping a tenant and then reusing its ID before servers
  from the previous record have fully shut down. This creates
  a possibility for a past server to start serving traffic
  for a freshly created tenant, which is unacceptable.

- we have a cache of tenant capabilites on every node. We do not yet
  implement an invalidation protocol to clear that cache upon tenant
  deletion. This makes it possible for requests from a new tenant
  to be authorized using the capabilities of a previous tenant, which
  is also unacceptable.

Until we implement the necessary synchronization, we can avoid the
correctness issues by preventing ID reuse.

This patch achieves this -- by preventing the deletion of rows in
system.tenants.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100613)
<!-- Reviewable:end -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100613)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @knz)

___
*[`pkg/multitenant/mtinfopb/info.go` line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100613#-NSCQTbf0Sxt9x-e-_uS:-NSCQTbg4Z_R3F-D_Pfg:b5x4bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8ff2114772af09559e773c50463729ef072755/pkg/multitenant/mtinfopb/info.go#L100)):*
> ```Go
> 	"ready":   DataStateReady,
> 	"drop":    DataStateDrop,
> 	"deleted": DataStateDeleted,
> ```

Are drop(ped) and delete(d) synonyms here? I think the name is confusing if `DataStateDrop` means that the drop/delete process started and `DataStateDeleted` means that the drop/delete process finished.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100613)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @ecwall)

___
*[`pkg/multitenant/mtinfopb/info.go` line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100613#-NSCQTbf0Sxt9x-e-_uS:-NSCRDYXAP5qN7O3rVmZ:b-wslx1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8ff2114772af09559e773c50463729ef072755/pkg/multitenant/mtinfopb/info.go#L100)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Are drop(ped) and delete(d) synonyms here? I think the name is confusing if `DataStateDrop` means that the drop/delete process started and `DataStateDeleted` means that the drop/delete process finished.
</blockquote></details>

yes "drop" means the data is in the process of being deleted; whereas "deleted" means this has completed. Feel free to propose better names.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100613)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @knz)

___
*[`pkg/multitenant/mtinfopb/info.go` line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100613#-NSCQTbf0Sxt9x-e-_uS:-NSCcPKZ6NoPnVPypTh-:bc7u9md) ([raw file](https://github.com/cockroachdb/cockroach/blob/db8ff2114772af09559e773c50463729ef072755/pkg/multitenant/mtinfopb/info.go#L100)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

yes "drop" means the data is in the process of being deleted; whereas "deleted" means this has completed. Feel free to propose better names.
</blockquote></details>

I don't care if we go with drop or deleted, but I think both should use the same term and the difference between the two should indicate if it is started or completed:
```
"dropping":    DataStateDropping,
"dropped":    DataStateDropped
```
or more explicitly
```
"drop started":    DataStateDropStarted,
"drop completed":    DataStateDropCompleted,
```


<!-- Sent from Reviewable.io -->

knz committed:
multitenant: display the dropping state as "dropping" not "drop"

This specific data state for secondary tenants indicate that the
tenant keyspace is queued for asynchronous deletion by a GC jobs;
however, it may not have been deleted yet.

Therefore, the visual representation of that state is best named
"dropping" instead of "drop".

Release note: None

knz commented:
Replaced with https://github.com/cockroachdb/cockroach/pull/101322
